unified-scheduler-pool/src/sleepless_testing.rs

```diff
@@ -0,0 +1,352 @@
+                            })
+                            .unwrap();
+                        true
+                    }
+                    // anchor is already observed.
+                    Less => false,
+                };
+
+            if should_change {
+                *current_check_point = anchored_check_point;
+                self.condvar.notify_all();
+            }
+        }
+
+        fn expected_next_index(&self, current_check_point: &String) -> usize {
+            let current_index = self
+                .check_points
+                .iter()
+                .position(|check_point| check_point == current_check_point)
+                .unwrap();
+            current_index.checked_add(1).unwrap()
+        }
+    }
+
+    lazy_static! {
+        static ref THREAD_REGISTRY: Mutex<HashMap<ThreadId, Arc<Progress>>> =
+            Mutex::new(HashMap::new());
+    }
+
+    #[must_use]
+    pub(crate) struct ActiveProgress(Arc<Progress>, ThreadId);
+
+    impl ActiveProgress {
+        fn new(progress: Arc<Progress>) -> Self {
+            let active_progress = Self(progress, current().id());
+            active_progress.activate();
+            active_progress
+        }
+
+        fn activate(&self) {
+            assert!(THREAD_REGISTRY
+                .lock()
+                .unwrap()
+                .insert(self.1, self.0.clone())
+                .is_none());
+        }
+
+        fn deactivate(&self) {
+            assert_eq!(
+                *self.0.check_points.last().unwrap(),
+                *self.0.current_check_point.lock().unwrap(),
+                "unfinished progress"
+            );
+            THREAD_REGISTRY.lock().unwrap().remove(&self.1).unwrap();
+        }
+    }
+
+    impl Drop for ActiveProgress {
+        fn drop(&mut self) {
+            self.deactivate();
+        }
+    }
+
+    /// Enable sleepless_testing with given check points being monitored, until the returned value
+    /// is dropped. This guarantees the check points are linearized in the exact order as
+    /// specified, among all of tracked threads.
+    pub(crate) fn setup(check_points: &[&dyn Debug]) -> ActiveProgress {
+        let progress = Arc::new(Progress::new(
+            check_points
+                .iter()
+                .map(|check_point| format!("{check_point:?}")),
+            current().name().unwrap_or_default().to_string(),
+        ));
+        ActiveProgress::new(progress)
+    }
+
+    /// Signal about the passage of the given check point. If sleepless_testing is enabled with the
+    /// check point monitored, this may block the current thread, not to violate the enforced order
+    /// of monitored check points.
+    pub(crate) fn at<T: Debug>(check_point: T) {
+        let mut registry = THREAD_REGISTRY.lock().unwrap();
+        if let Some(progress) = registry.get_mut(&current().id()).cloned() {
+            drop(registry);
```

Comment:
is this intentional? we drop the lock *before* the checkpoint is actually changed?

Reply:
yeah, i think dead lock is possible otherwise iirc